File: leetcode199.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        res = [root.val]
        layer = [root]
        while len(layer) > 0:
            tmp = []
            for x in layer:
                if x.left:
                    tmp.append(x.left)
                if x.right:
                    tmp.append(x.right)
            layer = tmp
            if len(layer) > 0:
                res.append(layer[-1].val)
        return res

    def rightSideView1(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        view = []
        def coll(root, depth):
            if root:
                if depth == len(view):
                    view.append(root.val)
                coll(root.right, depth+1)
                coll(root.left, depth+1)
        coll(root,0)
        return view

File: leetcodec++/leetcode199.cpp
//
// Created by 黄敏辉 on 2019-08-27.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> res;
        if(!root)
            return res;
        vector<TreeNode*> nodes;
        nodes.push_back(root);
        while(!nodes.empty()){
            vector<TreeNode*> tmp;
            res.push_back(nodes.end()->val)
            for(int i = 0; i < nodes.size(); i++){
                if (nodes[i]->left)
                    tmp.push_back(nodes[i]->left);
                if (nodes[i]->right)
                    tmp.push_back(nodes[i]->right);
            }
            nodes = tmp;
        }
        return res;
    }
};